File: Antlr/src/main/scala/io/github/chaosunity/casc/util/TypeResolver.scala
package io.github.chaosunity.casc.util

import io.github.chaosunity.antlr.CASCParser
import io.github.chaosunity.casc.parsing.`type`.{BuiltInType, ClassType, Type}
import org.apache.commons.lang3.{BooleanUtils, StringUtils}
import org.apache.commons.lang3.math.NumberUtils
import org.apache.commons.validator.routines.{DoubleValidator, FloatValidator, IntegerValidator}

object TypeResolver {
    def getFromTypeName(typeContext: CASCParser.TypeContext): Type = {
        if (typeContext == null) return BuiltInType.VOID

        val typeName = typeContext.getText
        val builtInType = getBuiltInType(typeName)

        if (typeName.equals("java.lang.String")) return BuiltInType.STRING

        if (builtInType.isDefined) return builtInType.get

        new ClassType(typeName)
    }

    def getFromValue(value: String): Type =
        if (StringUtils.isEmpty(value)) BuiltInType.VOID
        else if (NumberUtils.isCreatable(value)) {
            if (IntegerValidator.getInstance().validate(value) != null) BuiltInType.INT
            else if (FloatValidator.getInstance().validate(value) != null) BuiltInType.FLOAT
            else if (DoubleValidator.getInstance().validate(value) != null) BuiltInType.DOUBLE
            else throw new RuntimeException(s"$value is not a valid number syntax.")
        } else if (BooleanUtils.toBoolean(value)) BuiltInType.BOOLEAN
        else BuiltInType.STRING

    def getValueFromString(value: String, t: Type): Any =
        if (TypeChecker.isInteger(t)) value.toInt
        else if (TypeChecker.isFloat(t)) value.toFloat
        else if (TypeChecker.isDouble(t)) value.toDouble
        else if (TypeChecker.isBoolean(t)) value.toBoolean
        else if (t == BuiltInType.STRING) {
            var trimmedString = StringUtils.removeStart(value, "\"")
            trimmedString = StringUtils.removeEnd(trimmedString, "\"")

            trimmedString
        } else throw new RuntimeException("Object is not implemented yet.")

    private def getBuiltInType(typeName: String): Option[Type] =
        BuiltInType.enumSet.find(_.isAlias(typeName))
}

File: src/main/kotlin/io/github/chaosunity/casc/utils/TypeResolver.kt
package io.github.chaosunity.casc.utils

import io.github.chaosunity.casc.CASCParser
import io.github.chaosunity.casc.parsing.type.BuiltInType
import io.github.chaosunity.casc.parsing.type.ClassType
import io.github.chaosunity.casc.parsing.type.Type
import org.apache.commons.lang3.BooleanUtils
import org.apache.commons.lang3.math.NumberUtils

object TypeResolver {
    fun getFromTypeName(ctx: CASCParser.TypeContext?): Type {
        if (ctx == null) return BuiltInType.VOID

        val typeName = ctx.text

        if (typeName == "java.lang.String") return BuiltInType.STRING

        val builtInType = getBuiltInType(typeName)

        if (builtInType != null) return builtInType

        return ClassType(typeName)
    }

    fun getTypeByValue(value: String): Type {
        if (value.isEmpty()) return BuiltInType.VOID

        if (NumberUtils.isCreatable(value)) {
            when {
                value.toIntOrNull() != null -> return BuiltInType.INT
                value.toFloatOrNull() != null -> return BuiltInType.FLOAT
                value.toDoubleOrNull() != null -> return BuiltInType.DOUBLE
            }
        } else if (BooleanUtils.toBoolean(value)) return BuiltInType.BOOLEAN

        return BuiltInType.STRING
    }

    fun getValueByString(value: String, type: Type): Any =
        when {
            type.isInt() -> value.toInt()
            type.isFloat() -> value.toFloat()
            type.isDouble() -> value.toDouble()
            type.isBool() -> value.toBoolean()
            type.isString() -> value.removeSurrounding("\"", "\"")
            else -> throw RuntimeException("Object is not implemented yet.")
        }

    fun getBuiltInType(typeName: String): BuiltInType? =
        BuiltInType.values().find { it.typeName == typeName }
}